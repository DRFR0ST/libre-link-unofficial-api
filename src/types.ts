export enum LibreLinkUpEndpoints {
  Login = "llu/auth/login",
  Country = "llu/config/country?country=DE",
  Connections = "llu/connections",
}

/**
 * @description A response from the Libre Link Up API.
 */
export interface LibreResponse {
  status: number,
  data?: Record<string, any>;
  error?: Record<string, any>;
}

/**
 * @description The successful login response from the Libre Link Up API.
 */
export interface LibreLoginResponse extends LibreResponse {
  data: {
    /**
     * @description The user object from the Libre Link Up API.
     */
    user: LibreUser;
    /**
     * @description The messages from the Libre Link Up API. Usually only the count of unread messages.
     */
    messages: LibreDataMessages;
    /**
     * @description The notifications from the Libre Link Up API. Usually only the count of unresolved notifications.
     */
    notifications: LibreNotifications;
    /**
     * @description The authentication ticket from the Libre Link Up API. The authentication token is stored here.
     */
    authTicket: LibreAuthTicket;
    
    invitations: string[] | null;
    trustedDeviceToken: string | "";
  };
};

/**
 * @description The redirect response from the Libre Link Up API. Usually happens when attempting to log in with the wrong region.
 */
export interface LibreRedirectResponse extends LibreResponse {
  data: {
    redirect: boolean;
    region: string;
  };
};

/**
 * @description An error response from the Libre Link Up API.
 */
export interface LibreErrorResponse extends LibreResponse {
  error: {
    message: string,
  };
};

/**
 * @description A user object from the Libre Link Up API.
 */
export interface LibreUser {
  id: string;
  firstName: string;
  lastName: string;
  email: string;
  dateOfBirth: Date;
  /** Country of the user in alpha-2 format. eg. PL, DE, GB etc. */
  country: string;
  /** The language of the user interface in culture code format. eg. en-GB, de-DE, pl-PL etc. */
  uiLanguage: string;
  /** The language the system uses to communicate with the user (?) in culture code format. eg. en-GB, de-DE, pl-PL etc. */
  communicationLanguage: string;
  /** The account type of the user. For patient it's 'pat'. */
  accountType: string;
  created: Date;
  lastLogin: Date;
  uom: string;
  dateFormat: string;
  timeFormat: string;
  emailDay: number[];
  system: LibreSystem;
  details: Record<string, any>;
  programs: Record<string, any>;
  /** The practices that can access the user's data. */
  practices: Record<string, LibrePractice>;
  /** The devices that the user has connected to the account. */
  devices: Record<string, LibreDevice>;
  consents: LibreConsents;
}

interface LibreAuthTicket {
  token: string;
  expires: number;
  duration: number;
}

interface LibreDataMessages {
  unread: number;
}

interface LibreNotifications {
  unresolved: number;
}

interface LibreConsents {
  llu?: Llu;
  realWorldEvidence: RealWorldEvidence;
}

interface Llu {
  policyAccept: number;
  touAccept: number;
}

interface RealWorldEvidence {
  policyAccept: number;
  touAccept: number;
  history: { policyAccept: number, declined?: boolean }[];
}

interface LibreSystem {
  messages: LibreSystemMessages;
}

interface LibreSystemMessages {
  appReviewBanner: number;
  firstUsePhoenix: number;
  firstUsePhoenixReportsDataMerged: number;
  lluGettingStartedBanner: number;
  lluNewFeatureModal: number;
  lluOnboarding?: number;
  lvWebPostRelease: string;
  streamingTourMandatory: number;
}

/** The practice facility that can access the user's data. */
interface LibrePractice {
  id: string;
  practiceId: string;
  name: string;
  address1: string;
  city: string;
  state: string;
  zip: string;
  phoneNumber: string;
  records: null;
}

interface LibreDevice {
  id: string;
  nickname: string;
  sn: string;
  type: number;
  uploadDate: number;
}