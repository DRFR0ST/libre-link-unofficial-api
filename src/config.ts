
const LIBRE_LINK_API_URL = process?.env?.LIBRE_LINK_API_URL ?? "https://api-us.libreview.io";
const LIBRE_LINK_EMAIL = process?.env?.LIBRE_LINK_EMAIL;
const LIBRE_LINK_PASSWORD = process?.env?.LIBRE_LINK_PASSWORD;
const LIBRE_LINK_UP_VERSION = process?.env?.LIBRE_LINK_UP_VERSION ?? "4.7.0";
const LIBRE_LINK_PATIENT_ID = process?.env?.LIBRE_LINK_PATIENT_ID;
const VERBOSE = process?.env?.VERBOSE === "true";

/**
 * @description Configuration for the Libre Link Up API client.
 */
export const config = Object.freeze({
    apiUrl: LIBRE_LINK_API_URL,
    patientId: LIBRE_LINK_PATIENT_ID,
    credentials: {
        email: LIBRE_LINK_EMAIL,
        password: LIBRE_LINK_PASSWORD
    },
    lluVersion: LIBRE_LINK_UP_VERSION,
    verbose: VERBOSE
});