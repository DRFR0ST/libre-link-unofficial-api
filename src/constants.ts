import { Trend, TrendType } from "./types";

export const TREND_MAP: Trend[] = [
    Trend.NotComputable,
    Trend.SingleDown,
    Trend.FortyFiveDown,
    Trend.Flat,
    Trend.FortyFiveUp,
    Trend.SingleUp,
];

export const TREND_TYPE_MAP: TrendType[] = [
    "NotComputable",
    "SingleDown",
    "FortyFiveDown",
    "Flat",
    "FortyFiveUp",
    "SingleUp",
];