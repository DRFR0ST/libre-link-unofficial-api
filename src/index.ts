export { LibreLinkClient } from "./client";
export { GlucoseReading } from './reading';