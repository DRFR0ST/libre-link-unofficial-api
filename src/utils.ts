import { TREND_MAP } from "./constants";
import { RawGlucoseReading, LibreUser, GlucoseReading, MeasurementColor, LibreConnection, Trend } from "./types";

/**
 * Parse a Libre User object.
 * @param user The user object to parse.
 * @returns The parsed user object.
 */
export const parseUser = (user: Record<string, any>): LibreUser => ({
    ...(user as LibreUser),
    created: parseUnixTimestamp(user.created as unknown as number),
    lastLogin: parseUnixTimestamp(user.lastLogin as unknown as number),
    dateOfBirth: parseUnixTimestamp(user.dateOfBirth as unknown as number),
});

/**
 * @description Parse unix timestamps to Date objects.
 * @param timestamp The unix timestamp to parse.
 * @returns The parsed date.
 */
export const parseUnixTimestamp = (timestamp: number) => new Date(timestamp * 1000);

/**
 * @description Parse a glucose reading.
 * @param rawReading The raw glucose reading coming from the server to parse.
 * @param connection The connection object to use for parsing. Used for calculating isHigh and isLow.
 * @returns The parsed glucose reading.
 */
export const parseGlucoseReading = (rawReading: RawGlucoseReading, connection: LibreConnection): GlucoseReading => {
    // ! Calculates the isHigh and isLow properties based on the targetHigh and targetLow values. The two values coming from Libre Link Up seems to be incorrect.
    const isHigh = connection?.targetHigh < rawReading.Value;
    const isLow = connection?.targetLow > rawReading.Value;
   
    const parsedReading = Object.freeze({
        timestamp: new Date(rawReading.Timestamp),
        value: rawReading.Value,
        measurementColor: rawReading.MeasurementColor as MeasurementColor,
        isHigh,
        isLow,
        trend: getTrend(rawReading.TrendArrow)
    });

    return parsedReading;
}

/**
 * @description Sort by timestamp.
 * @param a The first item to compare.
 * @param b The second item to compare.
 * @returns The comparison result.
 * 
 * @example 
 * .sort(sortByTimestamp)
 */
export const sortByTimestamp = (a: GlucoseReading, b: GlucoseReading) => a.timestamp.getTime() - b.timestamp.getTime();

/**
 * @description Get the trend of a set of glucose readings.
 * @param trend The current trend.
 * @param defaultTrend The default trend to use if trend is undefined.
 */
const getTrend = (trend: number | undefined, defaultTrend: Trend = Trend.Flat) => TREND_MAP[trend!] ?? defaultTrend;