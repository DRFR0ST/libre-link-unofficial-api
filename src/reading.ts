import { TREND_TYPE_MAP } from "./constants";
import { MeasurementColor, RawGlucoseReading, Trend, TrendType } from "./types";
import { parseGlucoseReading } from "./utils";

const DEFAULT_OPTIONS = Object.freeze({
    targetHigh: 150,
    targetLow: 70
});

/**
 * @description A glucose reading class.
 */
export class GlucoseReading {
    /**
     * @description The timestamp of the glucose reading.
     */
    public timestamp: Date;
    /**
     * @description The value of the glucose reading in mg/dL.
     */
    public value: number;
    /**
     * @description The measurement color of the glucose reading. See {@link MeasurementColor}.
     */
    public measurementColor: MeasurementColor;
    /**
     * @description Whether the glucose reading is high, based on the patient's settings. Calculated by the library.
     */
    public isHigh: boolean;
    /**
     * @description Whether the glucose reading is low, based on the patient's settings. Calculated by the library.
     */
    public isLow: boolean;
    /**
     * @description The trend of the glucose reading. See {@link Trend}.
     */
    public trend: Trend;
  
    constructor(public _raw: RawGlucoseReading, public _options: { targetHigh: number, targetLow: number } = DEFAULT_OPTIONS) {
        const parsed = parseGlucoseReading(_raw, _options);

        this.value = parsed.value;
        this.timestamp = parsed.timestamp;
        this.measurementColor = parsed.measurementColor;
        this.isHigh = parsed.isHigh;
        this.isLow = parsed.isLow;
        this.trend = parsed.trend;
    }

    /**
     * @description The mmol value of the glucose reading.
     */
    get mmol() {
        return (this.value / 18).toFixed(1);
    }

    /**
     * @description The mg/dL value of the glucose reading.
     */
    get mgDl() {
        return this.value;
    }

    /**
     * @description The type of the trend. {@see TrendType}
     */
    get trendType(): TrendType {
        return TREND_TYPE_MAP[this.trend];
    }
}