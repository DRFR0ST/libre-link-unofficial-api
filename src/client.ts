import { config } from "./config";
import { GlucoseReading } from "./reading";
import { LibreLinkUpEndpoints, LibreLoginResponse, LibreResponse, LibreRedirectResponse, LibreUser, LibreConnection, LibreConnectionResponse, LibreLogbookResponse } from "./types";
import { encryptSha256, parseUser } from "./utils";

/**
 * A class for interacting with the Libre Link Up API.
 */
export class LibreLinkClient {
  private apiUrl = config.apiUrl;
  private accessToken: string | null = null;
  private patientId: string | null = config.patientId || null;
  private lluVersion: string | undefined;

  // A cache for storing fetched data.
  private cache = new Map<string, any>();

  constructor(private options: LibreLinkClientOptions = DEFAULT_OPTIONS) {
    if (!options?.email && !config.credentials.email)
      throw new Error("Libre Link Up credentials are missing.");

    if(options?.patientId)
      this.patientId = options.patientId;

    if(options?.lluVersion)
      this.lluVersion = options.lluVersion;

    // Merge the options with the default options.
    this.options = { ...DEFAULT_OPTIONS, ...options };
  }

  /**
   * @description Get the user data. Only available after logging in.
   */
  public get me(): LibreUser | null {
    if(!this.cache.has("user")) {
      console.warn("User data is not available. Please log in first.");
      return null;
    }

    return this.cache.get("user");
  }

  /**
   * @description Log into the Libre Link Up API using the provided credentials.
   */
  public async login(): Promise<LibreLoginResponse> {
    const email = this.options?.email || config.credentials.email;
    const password = this.options?.password || config.credentials.password;
    
    try {
      type LoginResponse = LibreLoginResponse | LibreRedirectResponse;
      
      // Attempt to login to the Libre Link Up API.
      const response = await this._fetcher<LoginResponse>(LibreLinkUpEndpoints.Login, {
        method: "POST",
        body: JSON.stringify({
          email,
          password
        }),
      });
  
      // If the status is 2, means the credentials are invalid.
      if(response.status === 2)
        throw new Error("Invalid credentials. Please ensure that the email and password work with the LibreLinkUp app.");

      if(!response.data) 
        throw new Error("No data returned from Libre Link Up API.");

      // If the response contains a redirect, update the region and try again.
      if("redirect" in response.data) {
        this.verbose("Redirecting to region:", response.data.region);
        const regionUrl = await this.findRegion(response.data.region);
        // Update the API URL with the region url.
        this.apiUrl = regionUrl;
        
        return await this.login();
      }

      // Set the access token for future requests.
      this.accessToken = response.data.authTicket?.token;

      // Cache the user data for future use. Log in again to refresh the user data.
      this.setCache("user", parseUser(response.data.user));

      this.verbose("Logged into Libre Link Up API.");

      return response as LibreLoginResponse;
    } catch(err) {
      const error = err as Error;

      console.error(error);
      throw new Error(`Error logging into Libre Link Up API. ${error.message}`);
    }
  }

  /**
   * @description Read the data from the Libre Link Up API.
   * @returns The latest glucose measurement from the Libre Link Up API.
   */
  public async read() {
    try {
      const response = await this.fetchReading();

      // Parse and return the latest glucose item from the response.
      return new GlucoseReading(response.data?.connection.glucoseItem, response.data.connection);
    } catch(err) {
      const error = err as Error;

      console.error(error);
      throw new Error(`Error reading data from Libre Link Up API. ${error.message}`);
    }
  }

  /**
   * @description Read the history data from the Libre Link Up API.
   */
  public async history() {
    try {
      const response = await this.fetchReading();

      const list = response.data.graphData.map((item) => new GlucoseReading(item, response.data.connection));

      return list;
    } catch(err) {
      const error = err as Error;

      console.error(error);
      throw new Error(`Error reading data from Libre Link Up API. ${error.message}`);
    }
  }

  /**
   * @description Read the logbook data from manual scans from the Libre Link Up API.
   */
  public async logbook() {
    try {
      const response = await this.fetchLogbook();

      const list = response.data.map((item) => new GlucoseReading(item));

      return list;
    } catch(err) {
      const error = err as Error;

      console.error(error);
      throw new Error(`Error reading data from Libre Link Up API. ${error.message}`);
    }
  }

  /**
     * @description Stream the readings from the Libre Link Up API.
     * @param intervalMs The interval between each reading. Default is 90 seconds.
     */
  public async *stream(intervalMs = 1000 * 90) {
    while (true) { // Keep streaming until manually stopped
      try {
        const reading = await this.read();
        yield reading;
        await new Promise(resolve => setTimeout(resolve, intervalMs));
      } catch (error) {
        console.error("Error fetching reading:", error); 
        throw error;
      }
    }
  }

  /**
   * @description Fetch the reading from the Libre Link Up API. Use to obtain the raw reading and more.
   * @returns The response from the Libre Link Up API.
   */
  public async fetchReading() {
    try {
      const patientId = await this.getPatientId();

      const headers = {
        "Account-Id": this.me?.id ? await encryptSha256(this.me.id) : "",
      };

      const response = await this._fetcher<LibreConnectionResponse>(`${LibreLinkUpEndpoints.Connections}/${patientId}/graph`, { headers });

      this.verbose("Fetched reading from Libre Link Up API.", JSON.stringify(response, null, 2));

      return response;
    } catch(err) {
      const error = err as Error;

      console.error(error);
      throw new Error(`Error fetching reading from Libre Link Up API. ${error.message}`);
    }
  }

  /**
   * @description Fetch the logbook from the Libre Link Up API. Use to obtain the list of manual scanned readings.
   * @returns The response from the Libre Link Up API.
   */
  public async fetchLogbook() {
    try {
      const patientId = await this.getPatientId();

      const headers = {
        "Account-Id": this.me?.id ? await encryptSha256(this.me.id) : "",
      };

      const response = await this._fetcher<LibreLogbookResponse>(`${LibreLinkUpEndpoints.Connections}/${patientId}/logbook`, { headers });

      this.verbose("Fetched logbook from Libre Link Up API.", JSON.stringify(response, null, 2));

      return response;
    } catch(err) {
      const error = err as Error;

      console.error(error);
      throw new Error(`Error fetching reading from Libre Link Up API. ${error.message}`);
    }
  }

  /**
   * @description Get the connections from the Libre Link Up API.
   */
  public async fetchConnections() {
    try {
      if(this.cache.has("connections"))
        return this.cache.get("connections");

      const headers = {
        "Account-Id": this.me?.id ? await encryptSha256(this.me.id) : "",
      };

      // Fetch the connections from the Libre Link Up API.
      const connections = await this._fetcher(LibreLinkUpEndpoints.Connections, { headers });

      this.verbose(`Fetched ${connections?.data?.length} connections from Libre Link Up API.`, JSON.stringify(connections, null, 2));

      if(!!connections?.data?.length)
        // Cache the connections for future use.
        this.setCache("connections", connections);

      return connections;
    } catch(err) {
      const error = err as Error;

      console.error(error);
      throw new Error(`Error fetching connections from Libre Link Up API. ${error.message}`);
    }
  }

  /**
   * @description Get the patient ID from the connections.
   */
  private async getPatientId() {
    const connections = await this.fetchConnections();

    // If there are no connections, throw an error.
    if(!connections.data?.length)
      throw new Error("No connections found. Please ensure that you have a connection with the LibreLinkUp app.");

    // Get the patient ID from the connections, or fallback to the first connection.

    let patientId = connections.data[0].patientId;

    if(this.patientId)
      connections.data.find(
        (connection: LibreConnection) => connection.patientId === this.patientId
      )?.patientId;

    if(!patientId)
      throw new Error(`Patient ID not found in connections. (${this.patientId})`);

    this.verbose("Using patient ID:", patientId);
    return patientId;
  }

  /**
   * @description Find the region in the Libre Link Up API. This is used when the API returns a redirect.
   * @param region The region to find.
   * @returns The server URL for the region.
   */
  private async findRegion(region: string) {
    try {
      const response = await this._fetcher(LibreLinkUpEndpoints.Country);

      // Find the region in the response.
      const lslApi = response.data?.regionalMap[region]?.lslApi;

      if(!lslApi)
        throw new Error("Region not found in Libre Link Up API.");

      return lslApi;
    } catch(err) {
      const error = err as Error;

      console.error(error);
      throw new Error(`Error finding region in Libre Link Up API. ${error.message}`);
    }
  }

  /**
   * @description A generic fetcher for the Libre Link Up API.
   * @param endpoint
   * @param options
   */
  private async _fetcher<T = LibreResponse>(endpoint: string, options: RequestInit = { headers: {} }) {
    const headers = new Headers({
      ...options.headers,
      Authorization: this.accessToken ? `Bearer ${this.accessToken}` : "",
  
      // Libre Link Up API headers
      product: 'llu.android',
      version: this.lluVersion || config.lluVersion,
  
      'accept-encoding': 'gzip',
      'cache-control': 'no-cache',
      connection: 'Keep-Alive',
      'content-type': 'application/json',
    });
    
    const requestOptions: FetchRequestInit = Object.freeze({
      ...options,
      headers
    });

    try {
      const response = await fetch(
        `${this.apiUrl}/${endpoint}`,
        requestOptions
      );

      this.verbose(
        `[${endpoint}] (${response.status})`,
        `Response from Libre Link Up API`,
        JSON.stringify(response, null, 2)
      );

      if (!response.ok) {
        const errorPayload = await response.json(); 
        const errorMessage = errorPayload?.message ?? JSON.stringify(errorPayload, null, 2)
        
        if(response.status === 429)
          throw new Error(`Too many requests. Please wait before trying again. ${errorMessage}`);

        throw new Error(
          `Error fetching data from Libre Link Up API with status ${response.status}. ${errorMessage}`
        );
      }

      const data = (await response.json()) as T;

      this.verbose(
        `[${endpoint}]`,
        `Data from Libre Link Up API`,
        JSON.stringify(data, null, 2)
      );

      return data;
    } catch (err) {
      const error = err as Error;

      console.error(error);
      throw new Error(`Error processing request to Libre Link Up API. ${error.message}`);
    }
  }

  /**
   * @description A verbose logger.
   * @param args
   */
  private verbose(...args: any[]) {
    if (config.verbose) console.log(...args);
  }

  /**
   * @description Cache a value, if caching is enabled.
   * @param key The key to cache the value under.
   * @param value The value to cache.
   */
  private setCache(key: string, value: any) {
    if(!this.options.cache) return;

    this.cache.set(key, value);
  }

  /**
   * @description Clear the cache.
   */
  public clearCache() {
    this.cache.clear();
  }
}

interface LibreLinkClientOptions {
  email?: string;
  password?: string;
  patientId?: string;
  cache?: boolean;
  lluVersion?: string;
}

const DEFAULT_OPTIONS: LibreLinkClientOptions = {
  cache: true,
};