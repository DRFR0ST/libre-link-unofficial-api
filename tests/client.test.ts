import { expect, test, beforeEach, describe } from "bun:test";
import { LibreLinkClient } from '../src';
import { mapObjectPropertiesToTypes } from "./utils";

describe('LibreLinkClient', () => {
  const client: LibreLinkClient = new LibreLinkClient();

  // Wait for a couple of seconds between tests to avoid 429 too many requests errors.
  beforeEach(() => new Promise((resolve) => setTimeout(resolve, 5000)));

  test('should be created', () => {
    expect(client).toBeTruthy();
  });

  test('should successfully login', async () => {
    await client.login();

    expect(client.me).toBeTruthy();
    expect(mapObjectPropertiesToTypes(client.me)).toMatchSnapshot();
  });

  test('should successfully fetch connections', async () => {
    const { data } = await client.fetchConnections();

    expect(data).toBeTruthy();
    expect(mapObjectPropertiesToTypes(data!)).toMatchSnapshot();
  });

  test('should successfully read data', async () => {
    const data = await client.read();

    expect(data).toBeTruthy();
    expect(mapObjectPropertiesToTypes(data!)).toMatchSnapshot();
  });

  test('should initialize with a patientId', async () => {
    const customClient = new LibreLinkClient({ patientId: "7f51ab27-c7c8-11ed-bcc3-0242ac110002" });

    await customClient.login();

    expect(customClient.me).toBeTruthy();
  });

  test('should throw error with an invalid patientId', async () => {
    const customClient = new LibreLinkClient({ patientId: "invalid-patient-id" });

    try {
      await customClient.login();
    } catch(err) {
      expect(err).toBeTruthy();
      expect(err.message).toMatch(/(Patient ID not found in connections. (invalid-patient-id))/i);
    }
  });
});