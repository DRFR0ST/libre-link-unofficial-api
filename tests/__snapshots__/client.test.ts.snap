// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`LibreLinkClient should successfully login 1`] = `
{
  "accountType": "string",
  "communicationLanguage": "string",
  "consents": {
    "realWorldEvidence": {
      "history": "array",
      "policyAccept": "number",
      "touAccept": "number",
    },
  },
  "country": "string",
  "created": {},
  "dateFormat": "string",
  "dateOfBirth": {},
  "details": {},
  "devices": {
    "a56c8d81-5eaa-11ee-9927-1639b8edd413": {
      "id": "string",
      "nickname": "string",
      "sn": "string",
      "type": "number",
      "uploadDate": "number",
    },
    "a96baf04-c7c8-11ed-8b8e-0242ac110006": {
      "id": "string",
      "nickname": "string",
      "sn": "string",
      "type": "number",
      "uploadDate": "number",
    },
    "c08364d5-f1ab-11ee-aaa5-5e24e2f78b68": {
      "id": "string",
      "nickname": "string",
      "sn": "string",
      "type": "number",
      "uploadDate": "number",
    },
  },
  "email": "string",
  "emailDay": "array",
  "firstName": "string",
  "id": "string",
  "lastLogin": {},
  "lastName": "string",
  "practices": {
    "5a68db16-0cda-49ed-9fab-68d087f2ac9c": {
      "address1": "string",
      "city": "string",
      "id": "string",
      "name": "string",
      "phoneNumber": "string",
      "practiceId": "string",
      "records": "null",
      "state": "string",
      "zip": "string",
    },
  },
  "programs": {},
  "system": {
    "messages": {
      "appReviewBanner": "number",
      "firstUsePhoenix": "number",
      "firstUsePhoenixReportsDataMerged": "number",
      "lluGettingStartedBanner": "number",
      "lluNewFeatureModal": "number",
      "lvWebPostRelease": "string",
      "streamingTourMandatory": "number",
    },
  },
  "timeFormat": "string",
  "twoFactor": {
    "primaryMethod": "string",
    "primaryValue": "string",
    "secondaryMethod": "string",
    "secondaryValue": "string",
  },
  "uiLanguage": "string",
  "uom": "string",
}
`;

exports[`LibreLinkClient should successfully fetch connections 1`] = `
{
  "0": {
    "alarmRules": {
      "c": "boolean",
      "f": {
        "d": "number",
        "th": "number",
        "thmm": "number",
        "tl": "number",
        "tlmm": "number",
      },
      "h": {
        "d": "number",
        "f": "number",
        "th": "number",
        "thmm": "number",
      },
      "l": {
        "d": "number",
        "on": "boolean",
        "th": "number",
        "thmm": "number",
        "tl": "number",
        "tlmm": "number",
      },
      "nd": {
        "i": "number",
        "l": "number",
        "r": "number",
      },
      "p": "number",
      "r": "number",
      "std": {},
    },
    "country": "string",
    "created": "number",
    "firstName": "string",
    "glucoseAlarm": "null",
    "glucoseItem": {
      "FactoryTimestamp": "string",
      "GlucoseUnits": "number",
      "MeasurementColor": "number",
      "Timestamp": "string",
      "TrendArrow": "number",
      "TrendMessage": "null",
      "Value": "number",
      "ValueInMgPerDl": "number",
      "isHigh": "boolean",
      "isLow": "boolean",
      "type": "number",
    },
    "glucoseMeasurement": {
      "FactoryTimestamp": "string",
      "GlucoseUnits": "number",
      "MeasurementColor": "number",
      "Timestamp": "string",
      "TrendArrow": "number",
      "TrendMessage": "null",
      "Value": "number",
      "ValueInMgPerDl": "number",
      "isHigh": "boolean",
      "isLow": "boolean",
      "type": "number",
    },
    "id": "string",
    "lastName": "string",
    "patientDevice": {
      "alarms": "boolean",
      "did": "string",
      "dtid": "number",
      "fixedLowAlarmValues": {
        "mgdl": "number",
        "mmoll": "number",
      },
      "fixedLowThreshold": "number",
      "hl": "number",
      "l": "boolean",
      "ll": "number",
      "u": "number",
      "v": "string",
    },
    "patientId": "string",
    "sensor": {
      "a": "number",
      "deviceId": "string",
      "lj": "boolean",
      "pt": "number",
      "s": "boolean",
      "sn": "string",
      "w": "number",
    },
    "status": "number",
    "targetHigh": "number",
    "targetLow": "number",
    "uom": "number",
  },
}
`;

exports[`LibreLinkClient should successfully read data 1`] = `
{
  "isHigh": "boolean",
  "isLow": "boolean",
  "measurementColor": "number",
  "timestamp": {},
  "value": "number",
}
`;
